src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -0,0 +1,225 @@
+            FMT_STRING(
+                "Order book is in a crossed state for {} - {} "
+                "asset pair.\nTop of the book is:\n\tAsk price: {}\n\tBid "
+                "price: {}\n\nWhere {} >= {}!"),
+            xdr::xdr_to_string(a), xdr::xdr_to_string(b), lowestAsk, highestBid,
+            highestBid, lowestAsk);
+    }
+    return {};
+}
+}
+
+void
+OrderBookIsNotCrossed::deleteFromOrderBook(OfferEntry const& oe)
+{
+    mOrderBook[{oe.selling, oe.buying}].erase(oe);
+}
+
+void
+OrderBookIsNotCrossed::addToOrderBook(OfferEntry const& oe)
+{
+    mOrderBook[{oe.selling, oe.buying}].emplace(oe);
+}
+
+void
+OrderBookIsNotCrossed::updateOrderBook(LedgerTxnDelta const& ltxd)
+{
+    for (auto const& entry : ltxd.entry)
+    {
+        // there are three possible "deltas" for an offer:
+        //      CREATED:  (nil)     -> LedgerKey
+        //      MODIFIED: LedgerKey -> LedgerKey
+        //      DELETED:  LedgerKey -> (nil)
+        if (entry.first.type() == InternalLedgerEntryType::LEDGER_ENTRY &&
+            entry.first.ledgerKey().type() == OFFER)
+        {
+            if (entry.second.previous)
+            {
+                deleteFromOrderBook(
+                    entry.second.previous->ledgerEntry().data.offer());
+            }
+            if (entry.second.current)
+            {
+                addToOrderBook(
+                    entry.second.current->ledgerEntry().data.offer());
+            }
+        }
+    }
+}
+
+std::string
+OrderBookIsNotCrossed::check(AssetPairSet const& assetPairs)
+{
+    if (mRollBackBeforeNextCheck)
+    {
+        mOrderBook = mRolledBackOrderBook;
+        mRollBackBeforeNextCheck = false;
+    }
+    for (auto const& assetPair : assetPairs)
+    {
+        auto checkCrossedResult =
+            checkCrossed(assetPair.first, assetPair.second, mOrderBook);
+        if (!checkCrossedResult.empty())
+        {
+            return checkCrossedResult;
+        }
+    }
+
+    return std::string{};
+}
+
+std::shared_ptr<Invariant>
+OrderBookIsNotCrossed::registerAndEnableInvariant(Application& app)
+{
+    auto invariant =
+        app.getInvariantManager().registerInvariant<OrderBookIsNotCrossed>();
+    app.getInvariantManager().enableInvariant(invariant->getName());
+    return invariant;
+}
+
+std::string
+OrderBookIsNotCrossed::getName() const
+{
+    return "OrderBookIsNotCrossed";
+}
+
+std::string
+OrderBookIsNotCrossed::checkOnOperationApply(Operation const& operation,
+                                             OperationResult const& result,
+                                             LedgerTxnDelta const& ltxDelta)
+{
+    updateOrderBook(ltxDelta);
+    auto assetPairs = extractAssetPairs(ltxDelta);
+    return assetPairs.size() > 0 ? check(assetPairs) : std::string{};
+}
+
+void
+OrderBookIsNotCrossed::snapshotForFuzzer()
+{
+    mRolledBackOrderBook = mOrderBook;
+}
+
+void
+OrderBookIsNotCrossed::resetForFuzzer()
+{
+    mRollBackBeforeNextCheck = true;
```

Comment:
I'm not 100% sure I understand the reasoning for this implementation. You could do `mOrderBook = mRolledBackOrderBook` here, right? I suppose I can imagine some reasons for not doing that. What was your reasoning? You probably want to put it in a comment if it's important.

Reply:
That was Nicolas's idea, and, I think, a good one: https://github.com/stellar/stellar-core/pull/2959#discussion_r592848513

The point is that some fuzz operation sets won't involve orders, and in those cases we call `OrderBookIsNotCrossed::checkOnOperationApply()` but not `OrderBookIsNotCrossed::check()`.  In such a case, we can skip the copy and allow the invariant's representation of the order book to be stale until we actually use it, in `OrderBookIsNotCrossed::check()`.  I'll add a comment to that effect.

Reply:
That's the same reason I came up with.